### DISCUSSION TITLE
Access username and password in custom phase script

### DISCUSSION THREAD

**Discussion by tiefenauer** [gitbls/sdm#65] [Q&A]
Could you explain in a bit more detail how the custom phase script works? I'm asking because I need the user name and password set with `--user` and `--password-user` in my custom phase script to set up Samba using `smbpasswd` and copying files into the user's home directory (for both tasks I need the user name).

Currently, I'm running SDM from my own bash script where I also configure user name and password and then passing them as custom parameters using the appropriate SDM switches:

```bash
user_name=...
user_password=...

sudo sdm --customize \
    --cscript custom-phase-script.sh \
    --custom1 $user_name \
    --custom2 $user_password \
    --plugin samba:smbconf=smb.conf \
```

I used the [sample file](https://github.com/gitbls/sdm/blob/master/sdm-customphase) for my `custom-phase-script.sh` and get them back in my script as follows:

```bash
user_name=$2
user_password=$3
```

However, I noticed there are some variables already pre-set, like `SDMPT`. Perhaps the user name and password are already 

The [Custom Phase Script Section in the Wiki](https://github.com/gitbls/sdm/wiki/Custom-Phase-Script) and the [Command Details Section](https://github.com/gitbls/sdm/wiki/Command-Details) do not give enough information about this.

**Comment by gitbls** [comment]
Did you review [this](https://github.com/gitbls/sdm/wiki/Example:-Custom-Phase-script)? Rather than me guessing at what would be helpful for you, would you please review it carefully and ask specific questions? I think this will be more productive than me guessing at what should be documented. Thx

**Reply by tiefenauer** [reply]
I did review the custom phase script (I even linked it in my description twice!). I tried to be as specific as I could, but I didn't reverse-engineer SDM.

Maybe these are more specific questions:
- Are there any other variables that are available in the custom phase script (besides `SDMPT`)?
- I'm currently passing the username and password set in `--user` and `--password-user` via custom params ((`--custom1` and `--custom2`) ). Is there an easier way to access them in the custom phase script?
- How can I access custom variables (`--custom[1-4]`) in the custom phase script?

**Reply by tiefenauer** [reply]
Here's the full script I'm using to create and burn a customized image using SDM: https://github.com/tiefenauer/mironibox/blob/main/flash.sh

Hope that helps

**Comment by gitbls** [comment]
Ah, OK, now I understand the script you initially included. I was very confused by the custom username and passwords. Got it.

At the moment the list of what variables are available is in /etc/sdm/cparams in an sdm-customized img. You can also see the list in sdm-cparams, function writeconfig.

But, there's no clear switch-to-variable information available. I'll add this to the wiki, and also document the environment differences between phase 0 (mounted disk) and phase 1 (container) from a code perspective.

Looking forward to your feedback on it after it's completed.

So, to your specific questions

- Yes, there are. That is what I mentioned above. There is one other env var, SDMNSPAWN, that I've been meaning to see if I can eliminate it.
- `--user` is the variable `$myuser`, and `--password-user` is `$passworduser`
- `--custom[1-4]` are `$custom[1-4]`

**Comment by gitbls** [comment]
I've posted a new wiki page: [Programming plugins and custom phase scripts](https://github.com/gitbls/sdm/wiki/Programming-plugins-and-custom-phase-scripts) which should address your questions. Appreciate you having a look at it and providing feedback.

**Reply by tiefenauer** [reply]
Nice, just what I needed. Thanks for updating the docs, I appreciate it. :heart: 
Just one thing: The switches, where they can be used (customization or burning), what they do and how their values can be accessed in the phases are now distributed over several wiki pages. For your own sake it might be worth keeping all the information in one place and reference it where needed. It will make maintaining the docs easier.

If you keep the docs under version control I (and other users) could help keeping it up to date by opening pull requests. I don't think there's a way to do this with the Wiki... Up to you!

**Comment by gitbls** [comment]
I think you're correct about the wiki. In general, the Wiki is a good approach, since there's so much information now. It used to all be on one page, which was a definite lose as the doc requirements grew.

I've been thinking about keeping the wiki sources on the github, and doing wiki updates via copy/paste when the sources get updated.

Maybe I'll get a break from adding features to sdm, and my "other project" pile will stop growing so I can attend to the wiki 🤣

**Comment by gitbls** [comment]
I've moved the Wiki sources into the Docs subdirectory. If you clone the repo the docs can be browsed locally by installing a Markdown Viewer extension. The Wiki will be evaporated soon.
